base/dates/periods.jl

```diff
@@ -166,10 +166,55 @@ type CompoundPeriod <: AbstractTime
                     break
                 end
             end
-            return new(reverse!(pc))
+            p = reverse!(pc)
+            n = length(p)
         else
             return new(resize!(p, n))
         end
+        # reduce the amount of mixed positive/negative Periods.
+        if n > 0
+            pc = sizehint!(Period[], n)
+            i = n
+            while i > 0
```

Comment:
So the canonicalization needs O(n^2) computation, due to the nested `i` and `j` loops?  I guess `n` is never going to be more than 10 or so, but still this is surprising.